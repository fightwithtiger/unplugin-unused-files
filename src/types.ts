export type TargetType = string | RegExp | Array<string | RegExp>
export interface Options {
  // define your plugin options here
  target?: TargetType
  excludes?: string[]
}
