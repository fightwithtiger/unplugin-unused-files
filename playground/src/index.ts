import { add } from 'util/math'
import dayjs from 'dayjs'
import { mius } from './util'

export function getContent() {
  const day = dayjs('2019-01-25').format('DD/MM/YYYY')
  return `${add(1, 1)}  ${day} ${mius(2, 5)}`
}
