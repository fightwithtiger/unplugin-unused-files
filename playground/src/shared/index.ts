export function isNumber(a: unknown) {
  return typeof a === 'number'
}
