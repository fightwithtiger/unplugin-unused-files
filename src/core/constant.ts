import { resolve, sep } from 'path'

export const SYSTEM_SEP = sep

export const ROOT = resolve('.', './src')
