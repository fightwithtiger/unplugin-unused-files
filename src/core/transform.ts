import fs from 'fs'
import { resolve } from 'path'
import type { Options, TargetType } from '../types'
import { ROOT, SYSTEM_SEP } from './constant'

const loadedFiles: string[] = []
let targetRegs: RegExp[]

export function getUnusedFiles() {
  const allFiles = getAllFiles()
  const processedFiles = allFiles.filter((file) => {
    let flag = false
    for (const reg of targetRegs) {
      if (reg.test(file)) {
        flag = true
        break
      }
    }
    return flag
  })
  return processedFiles.filter(file => !loadedFiles.includes(file))
}

export function getAllFiles(dir: string = ROOT) {
  try {
    const allFiles: string[] = []
    const files = fs.readdirSync(dir)
    files.forEach((file: any) => {
      const pth = resolve(dir, file)
      const stat = fs.statSync(pth)
      if (stat.isDirectory()) {
        const more = getAllFiles(pth)
        allFiles.push(...more)
      } else {
        allFiles.push(pth)
      }
    })

    return allFiles
  } catch (err) {
    console.error(err)
    return []
  }
}

export function mark(id: string, options: Options | undefined) {
  const { target = 'src' } = options || {} as Options
  targetRegs = getTargetRegs(target)
  const path = fixPath(id)
  for (const reg of targetRegs) {
    if (reg.test(path)) {
      loadedFiles.push(path)
      break
    }
  }

  return !!path
}

function fixPath(path: string) {
  return path.replace(/[\\/]/g, SYSTEM_SEP)
}

function getTargetRegs(target: TargetType) {
  const targetRegs = []
  if (typeof target === 'string') {
    const path = fixPath(target)
    targetRegs.push(new RegExp(path))
  } else if (Array.isArray(target)) {
    target.forEach((item) => {
      targetRegs.push(...getTargetRegs(item))
    })
  } else {
    targetRegs.push(target)
  }
  return targetRegs
}
