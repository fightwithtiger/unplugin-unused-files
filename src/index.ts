import { createUnplugin } from 'unplugin'
import { getAllFiles, getLoadedFiles, getUnusedFiles, mark } from './core/index'
import type { Options } from './types'

export default createUnplugin<Options | undefined>(options => ({
  name: 'unplugin-unused-files',
  transformInclude(id) {
    return mark(id, options)
  },
  transform(code) {
    return code
  },
  buildEnd() {
    const unusedFiles = getUnusedFiles()

    console.log('\nloadedFiles', unusedFiles)
  },
}))
