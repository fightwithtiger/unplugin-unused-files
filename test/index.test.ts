import { describe, expect, it } from 'vitest'
import { getAllFiles } from '../src/core'

describe('getAllFiles', () => {
  it('src', () => {
    const files = getAllFiles('src')
    expect(files).toMatchSnapshot()
  })
})
