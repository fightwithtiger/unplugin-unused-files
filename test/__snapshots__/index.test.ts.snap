// Vitest Snapshot v1

exports[`getAllFiles > src 1`] = `
[
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\core\\\\constant.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\core\\\\index.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\core\\\\transform.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\esbuild.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\index.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\nuxt.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\rollup.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\types.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\vite.ts",
  "E:\\\\work\\\\project\\\\published\\\\unplugin-starter\\\\src\\\\webpack.ts",
]
`;
